clearwater-infrastructure/usr/share/clearwater/bin/clearwater-check-config

```diff
@@ -0,0 +1,165 @@
+#!/bin/bash
+
+# @file check-clearwater-config
+#
+# Copyright (C) Metaswitch Networks 2014
+# If license terms are provided to you in a COPYING file in the root directory
+# of the source code repository by which you are accessing this code, then
+# the license outlined in that COPYING file applies to your use.
+# Otherwise no rights are granted except for those provided to you by
+# Metaswitch Networks in a separate written agreement.
+
+CONFIG_DIR=/etc/clearwater
+
+LOCAL_CONFIG=$CONFIG_DIR/local_config
+SHARED_CONFIG=$CONFIG_DIR/shared_config
+USER_SETTINGS=$CONFIG_DIR/user_settings
+CLEARWATER_CONFIG=$CONFIG_DIR/config
+
+# Utility function for writing messages to stderr.
+error()
+{
+  echo "ERROR: $@" >&2
+}
+
+# Check that an option is not in a specific file. If it is, print an error to
+# stderr and return a false return code.
+# Params:
+#   1  - The file to search
+#   2+ - The options to check
+check_options_not_in_file()
+{
+  # This function is split up into two parts: check_options_not_in_file which is
+  # the entry point and int_check_options_not_in_file which actually does the
+  # work. The split is because the latter sources in config files and will
+  # pollute the namespace so we use the former function to run the latter in a
+  # subshell. We could just do this all in one function but it breaks syntax
+  # highlighting horribly :-(.
+  ( int_check_options_not_in_file $@ )
+}
+
+int_check_options_not_in_file()
+{
+  local local_rc=0
+
+  filename=$1;
+  shift
+
+  source $filename
+
+  for opt_name in $@; do
+    # Check if the option is set. The `${!opt}` is bash indirection - it
+    # retrieves the value of the parameter whose name is the value of `opt`
+    if [ -n "${!opt_name}" ]; then
+      error "$opt_name should not be placed in $filename"
+      local_rc=1
+    fi
+  done
+
+  return $local_rc
+}
+
+# Check that a config file exists and is syntactically and semantically valid.
+# Params:
+#   1 - Either "mandatory" (if the file is required) or "optional" (otherwise).
+#   2 - The full name and path of the config file to check.
+check_file()
+{
+  local required=$1
+  local filename=$2
+
+  # Check that the file exists.
+  if [[ ! -e "$filename" ]]; then
+    if [[ "$required" = mandatory ]]; then
+      error "$filename does not exist"
+      return 1
+    fi
+    return 0
+  fi
+
+  # There is no need to check that the file has read permissions - all files are
+  # readable to root.
+
+  # Check that the file is syntactically valid.
+  if ! bash -n $filename; then
+    error "$filename is not syntactically valid (see previous errors for details)"
+    return 1
+  fi
+
+  return 0
+}
+
+# Check that certain important config options are not put in the wrong place.
+check_option_locations()
+{
+  local local_rc=0
+
+  # First, check that there are no options in shared config that really
+  # shouldn't be shared. log_level is included in this list as it only works
+  # properly when its put in user_settings.
```

Comment:
*it's. Also, I don't think log_level should be in this list - we should be fixing up any cases that require it to be in user_settings.

Reply:
Agree - fixed.